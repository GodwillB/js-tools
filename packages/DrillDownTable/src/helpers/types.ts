import { Dictionary } from '@onaio/utils';
import { TableJSXProps } from '..';

export type DrillDownFilter<D extends object> = (
  props: DrillDownTableProps<D>,
  parentId: string
) => D[];

/** describes props for the DrillDownTable component */
export interface DrillDownTableProps<D extends object>
  extends Omit<TableJSXProps<D>, 'fetchData' | 'parentNodes'> {
  extraCellProps?: Dictionary /** props to be given to CellComponent */;
  CellComponent: React.ElementType /** The component used to render the cell that has the drill down */;
  loading: boolean /** if loading */;
  loadingComponent: React.ElementType /** custom component to show whilst loading is true */;
  drillDownFilter: DrillDownFilter<D>;
}
