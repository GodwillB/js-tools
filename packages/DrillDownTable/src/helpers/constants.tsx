/** Constants */
import React from 'react';

export const ID = 'id';
export type ID = typeof ID;

export const PARENT_ID = 'parent_id';
export type PARENT_ID = typeof PARENT_ID;

export const ROOT_PARENT_ID = null;
export type ROOT_PARENT_ID = typeof ROOT_PARENT_ID;

export const LINKER_ITEM_CSS_CLASS = 'dd-linker-item';
export type LINKER_ITEM_CSS_CLASS = typeof LINKER_ITEM_CSS_CLASS;

export const CLICKABLE_CSS_CLASS = 'dd-linker-item dd-clickable';
export type CLICKABLE_CSS_CLASS = typeof CLICKABLE_CSS_CLASS;

export const CARET_CSS_CLASS = 'dd-caret';
export type CARET_CSS_CLASS = typeof CARET_CSS_CLASS;

export const CARET_SPAN = <span className={CARET_CSS_CLASS}>&nbsp;&#9660;</span>;
export type CARET_SPAN = typeof CARET_SPAN;

export const DEFAULT_ROW_HEIGHT = '1.5em';
export type DEFAULT_ROW_HEIGHT = typeof DEFAULT_ROW_HEIGHT;

export const NO_DATA_FOUND = 'No Data Found';
export type NO_DATA_FOUND = typeof NO_DATA_FOUND;

export const NEXT = 'Next';
export type NEXT = typeof NEXT;

export const OF = 'Of';
export type OF = typeof OF;

export const PAGE = 'Page';
export type PAGE = typeof PAGE;

export const PREVIOUS = 'Previous';
export type PREVIOUS = typeof PREVIOUS;

export const ROWS_TO_DISPLAY = 'Rows to display';
export type ROWS_TO_DISPLAY = typeof ROWS_TO_DISPLAY;

export const LOADING = 'Loading';
export type LOADING = typeof LOADING;

export const PAGE_SIZE_CATEGORIES = [10, 20, 30, 50];
export type PAGE_SIZE_CATEGORIES = typeof PAGE_SIZE_CATEGORIES;
