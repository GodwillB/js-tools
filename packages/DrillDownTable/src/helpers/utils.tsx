import { DrillDownColumn } from '../components/TableJSX';

/** Takes an object and returns columns that are compatible with react-table
 * derived from the object's keys
 */
export function columnsFromObjects<D extends object>(items: D[]): Array<DrillDownColumn<D>> {
  if (items.length < 1) {
    return [];
  }
  return (Object.keys(items[0]).map(k => ({ Header: k, accessor: k })) as unknown) as Array<
    DrillDownColumn<D>
  >;
}
