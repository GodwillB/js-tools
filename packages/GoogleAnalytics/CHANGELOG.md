# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.0.2](https://github.com/onaio/js-tools/compare/@onaio/google-analytics@0.0.1...@onaio/google-analytics@0.0.2) (2020-03-19)

**Note:** Version bump only for package @onaio/google-analytics

## 0.0.1 (2020-03-19)

**Note:** Version bump only for package @onaio/google-analytics
