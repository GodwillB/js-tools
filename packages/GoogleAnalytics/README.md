# GoogleAnalytics

GoogleAnalytics works with [React ga](https://github.com/react-ga/react-ga) to include Google Analytics code in a website or app that uses [React](https://facebook.github.io/react/)

## Installation

```sh

yarn add @onaio/google-analytics
```

## Tracking page views

A couple of approaches are available for use

### WithGATracker

A high order component `WithGATracker` wraps an entire `App` or `Route` component

#### Usage

```
import { WithGATracker, initGoogleAnalytics,setDimensions } from '@onaio/google-analytics'

const initializeOptions  = {
    testMode: true
}
initGoogleAnalytics('UA-000000-01', initializeOptions)

class App extends Component {
    ....

    render() {
        const dimensions = {
            env: 'test',
            username: 'logged-in-username'
        }
        setDimensions(dimensions);
        return (
            ....
           <Route exact path="/users" component={WithGATracker(Users)} />
        )
    }
}
```

This approach however forces components (and all of their children components) to unmount and remount every time the URL changes.

### RouterTracker (Best approach)

Uses a `Route` that matches everything so that it can be re-rendered on every route change.

#### Usage

```
import { RouteTracker, initGoogleAnalytics } from '@onaio/google-analytics'
....

const initializeOptions  = {
    testMode: true
}
initGoogleAnalytics('UA-000000-01', initializeOptions)

class App extends Component {
    ...
    render () {
        const dimensions = {
            env: 'test',
            username: 'logged-in-username'
        }
        setDimensions(dimensions);
        return (
            <BrowserRouter>

                { <RouteTracker /> }

                <Switch>
                    <AuthorizedRoute path='/' component={Home} />
                </Switch>
            </BrowserRouter>
        )
    }

}
```

Note that the `RouteTracker` is not wrapped inside the `Switch`

### API

Initialization must be done before any of the other tracking functions.

#### initGoogleAnalytics(trackingCode, initializeOptions, dimensions)

Returns a boolean. `true` if the initialization was a success, `false` otherwise.

Takes the following arguments:

1. `trackingCode` **(Required)**: A string value for the Google analytics tracking code.

2. `initializeOptions` **(Optional)**: Initialization options as defined by `[react-ga](https://www.npmjs.com/package/react-ga#reactgainitializegatrackingid-options)`. The default options if not provided are:

```
{
  testMode: false
};
```

The `InitializeOptions` interface looks like this

```ts
export interface InitializeOptions {
  debug?: boolean;
  gaAddress?: string;
  testMode?: boolean;
  titleCase?: boolean;
  gaOptions?: GaOptions;
  alwaysSendToDefaultTracker?: boolean;
  standardImplementation?: boolean;
}
```

#### setDimensions(dimensions)

Set [custom dimensions](https://www.npmjs.com/package/react-ga#reactgasetfieldsobject) (user attributes) to be tracked.

Takes the following arguments

1. `dimensions` **(Required)**:

The `Dimensions` interface looks like this

```ts
import { FieldsObject } from 'react-ga';

export interface Dimensions extends FieldsObject {
  env?: string;
  username?: string;
}
```
