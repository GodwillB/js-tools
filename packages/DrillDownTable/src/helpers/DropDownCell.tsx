import React from 'react';
import { CARET, CLICKABLE_CSS_CLASS, LINKER_ITEM_CSS_CLASS } from './constants';

export interface DropDownCellProps {
  cellValue: Node;
  hasChildren: boolean;
}

const DropDownCell: React.ElementType = (props: DropDownCellProps) => {
  const { cellValue, hasChildren } = props;
  return (
    <div className={hasChildren ? CLICKABLE_CSS_CLASS : LINKER_ITEM_CSS_CLASS}>
      <span>
        {cellValue}
        {hasChildren && CARET}
      </span>
    </div>
  );
};

export default DropDownCell;
