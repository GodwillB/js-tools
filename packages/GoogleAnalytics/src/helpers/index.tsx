import memoize from 'memoize-one';
import GoogleAnalytics, { FieldsObject, InitializeOptions } from 'react-ga';

/**
 * Interface defining the visitor attributes been tracked
 */
export interface Dimensions extends FieldsObject {
  env?: string;
  username?: string;
}

/**
 * helper function to execute the page view Google Analytics tracking
 * @param {string} page the url string of the page view being tracked
 */
export const trackPage = (page: string): void => {
  GoogleAnalytics.pageview(page);
};

/**
 * Default values for react ga initialization options
 */
export const defaultInitializeOptions: InitializeOptions = {
  testMode: false
};

/**
 * Helper function to initialize Google Analytics
 * @param {string} trackingCode Google analytics tracking code
 * @param {InitializeOptions} initializeOptions React GA initialization options
 * @returns {boolean} True if initialization was successful, false otherwise
 */
export const initGoogleAnalytics = memoize(
  (
    trackingCode: string,
    initializeOptions: InitializeOptions = defaultInitializeOptions
  ): boolean => {
    const isGAEnabled = !!(trackingCode && trackingCode.length);

    if (isGAEnabled) {
      GoogleAnalytics.initialize(trackingCode, initializeOptions);
    }

    return isGAEnabled;
  }
);

/**
 * Helper function to set custom dimensions
 * @param {Dimensions} dimensions User attributes to track
 */
export const setDimensions = memoize((dimensions: Dimensions): void => {
  GoogleAnalytics.set(dimensions);
});
