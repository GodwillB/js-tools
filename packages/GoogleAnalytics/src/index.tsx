import { RouteTracker } from './components/GARoute';
import WithGATracker from './components/WithGATracker';
import { Dimensions, initGoogleAnalytics, setDimensions } from './helpers';

export { WithGATracker, RouteTracker, initGoogleAnalytics, Dimensions, setDimensions };
