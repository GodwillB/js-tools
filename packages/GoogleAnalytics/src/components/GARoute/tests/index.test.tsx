import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { createBrowserHistory, Location } from 'history';
import React from 'react';
import GoogleAnalytics from 'react-ga';
import { Router } from 'react-router';
import RouteTracker, { GARoute } from '..';
import { Dimensions } from './../../../helpers';

const history = createBrowserHistory();

describe('/components/GARoute', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  const dimensions: Dimensions = {
    env: 'test',
    username: 'superman'
  };
  const location: Location = {
    hash: undefined,
    pathname: '/home',
    search: '?foo=bar',
    state: undefined
  };
  const props = {
    dimensions,
    history,
    location,
    match: {
      isExact: true,
      params: {},
      path: '/home',
      url: '/home'
    }
  };

  it('renders correctly without crashing', () => {
    GoogleAnalytics.pageview = jest.fn();
    const wrapper = mount(<GARoute {...props} />);
    expect(wrapper.isEmptyRender()).toBe(true);
    expect(GoogleAnalytics.pageview).toBeCalledWith('/home?foo=bar');
    wrapper.unmount();
  });

  it('tracks pageview if the component did not unmount and URL changes', () => {
    GoogleAnalytics.pageview = jest.fn();
    const wrapper = mount(<GARoute {...props} />);
    wrapper.setProps({
      location: {
        ...props.location,
        search: '?foo=baz'
      }
    });
    expect(GoogleAnalytics.pageview).toBeCalledWith('/home?foo=baz');
  });

  it('renders the route tracker correctly', () => {
    const wrapper = mount(
      <Router history={history}>
        <RouteTracker />
      </Router>
    );
    expect(toJson(wrapper.find('RouteTracker'))).toMatchSnapshot();
  });
});
