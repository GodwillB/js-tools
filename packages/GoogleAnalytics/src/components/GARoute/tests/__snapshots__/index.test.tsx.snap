// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/components/GARoute renders the route tracker correctly 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <RouteTracker
    dimensions={
      Object {
        "env": "test",
        "username": "superman",
      }
    }
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
    location={
      Object {
        "hash": undefined,
        "pathname": "/home",
        "search": "?foo=bar",
        "state": undefined,
      }
    }
    match={
      Object {
        "isExact": true,
        "params": Object {},
        "path": "/home",
        "url": "/home",
      }
    }
  >
    <Route
      component={[Function]}
    >
      <GARoute
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "createHref": [Function],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        }
        location={
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          }
        }
        match={
          Object {
            "isExact": true,
            "params": Object {},
            "path": "/",
            "url": "/",
          }
        }
      />
    </Route>
  </RouteTracker>
</Router>
`;
